### PR TITLE
Adding forecast capabilities for MLR and Renewal equation model

### PR DESCRIPTION
**Pull request by marlinfiggins**
### Adding a forecast time window and exporting forecasts

We have yet to actually forecast anything. Currently, we're implementing nowcasts for the renewal equation model and nothing for MLR if I understand correctly.

This PR explicitly adds a forecast time length to the MLR and renewal models which can be loaded from the model config. Currently, I'm using a default of 2 weeks.

This PR will only work using the current master branch of evofr though, so I will believe I will need to make a new evofr release.

### CODE REVIEW

**Comment by joverlee521** [comment]
Waiting on a new release of evofr, will need to update the workflows [docker-image tag](https://github.com/nextstrain/forecasts-ncov/blob/3e1ef3f51c5630093421f1f9a68f520dd5ff5d84/.github/workflows/run-models.yaml#L64).

**Comment by marlinfiggins** [comment]
New evofr version has been released 0.1.17!

**Comment by joverlee521** [comment]
> New evofr version has been released 0.1.17!

Awesome @marlinfiggins! I've triggered the [CI action](https://github.com/nextstrain/docker-base/actions/runs/4622997842) for Nextstrain docker-base to pull in the latest version.

**Comment by joverlee521** [comment]
I just triggered a [test run](https://github.com/nextstrain/forecasts-ncov/actions/runs/4633211396) of the models with this branch since the docker-base has been successfully updated.

**Review by joverlee521 — CHANGES_REQUESTED** [review]
@marlinfiggins this is raising some errors when I'm testing locally with: 
```
nextstrain build . --configfile config/config.yaml --config data_provenances=open
```

For the run-renewal-model script:
```
Traceback (most recent call last):
  File "/nextstrain/build/./scripts/run-renewal-model.py", line 373, in <module>
    multi_posterior = fit_models(
  File "/nextstrain/build/./scripts/run-renewal-model.py", line 270, in fit_models
    posterior = inference_method.fit(model, data, name=location)
  File "/nextstrain/build/./scripts/run-renewal-model.py", line 122, in fit
    init_strat, _ = ef.init_to_MAP(model, data, iters=30_000, lr=self.lr)
  File "/usr/local/lib/python3.10/site-packages/evofr/infer/InferSVI.py", line 114, in init_to_MAP
    MAP = infer_map.fit(model, data)
  File "/usr/local/lib/python3.10/site-packages/evofr/infer/InferSVI.py", line 77, in fit
    self.handler.fit(model.model_fn, guide, input, self.iters)
  File "/usr/local/lib/python3.10/site-packages/evofr/infer/SVI_handler.py", line 51, in fit
    self.init_svi(model, guide, data)
  File "/usr/local/lib/python3.10/site-packages/evofr/infer/SVI_handler.py", line 43, in init_svi
    svi_state = self.svi.init(self.rng_key, **data)
  File "/usr/local/lib/python3.10/site-packages/numpyro/infer/svi.py", line 180, in init
    guide_trace = trace(guide_init).get_trace(*args, **kwargs, **self.static_kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/handlers.py", line 171, in get_trace
    self(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/primitives.py", line 105, in __call__
    return self.fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/primitives.py", line 105, in __call__
    return self.fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/infer/autoguide.py", line 430, in __call__
    self._setup_prototype(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/infer/autoguide.py", line 404, in _setup_prototype
    super()._setup_prototype(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/infer/autoguide.py", line 154, in _setup_prototype
    ) = initialize_model(
  File "/usr/local/lib/python3.10/site-packages/numpyro/infer/util.py", line 607, in initialize_model
    ) = _get_model_transforms(substituted_model, model_args, model_kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/infer/util.py", line 405, in _get_model_transforms
    model_trace = trace(model).get_trace(*model_args, **model_kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/handlers.py", line 171, in get_trace
    self(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/primitives.py", line 105, in __call__
    return self.fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/primitives.py", line 105, in __call__
    return self.fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/numpyro/primitives.py", line 105, in __call__
    return self.fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/evofr/models/renewal_model/model_factories.py", line 41, in _renewal_model
    v_idx = [
  File "/usr/local/lib/python3.10/site-packages/evofr/models/renewal_model/model_factories.py", line 42, in <listcomp>
    var_names.index(s) for s in gen_v_names
ValueError: 'Delta' is not in list
```


For the run-mlr-model script:
```
Traceback (most recent call last):
  File "/nextstrain/build/./scripts/run-mlr-model.py", line 286, in <module>
    export_results(multi_posterior, ps, export_path, data_name)
  File "/nextstrain/build/./scripts/run-mlr-model.py", line 199, in export_results
    ef.posterior.get_sites_variants_tidy(
  File "/usr/local/lib/python3.10/site-packages/evofr/posterior/posterior_helpers.py", line 428, in get_sites_variants_tidy
    entries.extend(tidy_site(site))
  File "/usr/local/lib/python3.10/site-packages/evofr/posterior/posterior_helpers.py", line 350, in tidy_site_date
    entry_med["value"] = np.around(med[index, v], decimals=3)
IndexError: index 30 is out of bounds for axis 0 with size 30
```

**Comment by marlinfiggins** [comment]
TL;DR: Two PRs in Evofr should fix this issue. They're included in 0.1.18 which has been released. Code has been updated to take advantage of these changes.

First thing, I'm noticed is that Renewal equation error is caused by a bug in Evofr which is being fixed by this [PR](https://github.com/blab/evofr/pull/21). I thought we were eliminating the renewal based forecasts though I guess it is good to fix this regardless.

Secondly, I found the issue with the MLR model. It was that it was trying to date the forecasts using the dates from the original data. I was able to fix this with this [PR](https://github.com/blab/evofr/pull/22) in evofr.

I've updated the two export lines to indicate which variables are forecasts and which are not.

**Comment by joverlee521** [comment]
Triggered [action](https://github.com/nextstrain/docker-base/actions/runs/4634197507) for Nextstrain docker-base to pull in evofr 0.1.18

**Review by joverlee521 — APPROVED** [review]
The [test run](https://github.com/nextstrain/forecasts-ncov/actions/runs/4640875626) completed successfully and uploaded results to:
* `s3://nextstrain-data/files/workflows/forecasts-ncov/gisaid/trial/forecasting-time-length/nextstrain_clades/global/renewal/latest_results.json`
* `s3://nextstrain-data/files/workflows/forecasts-ncov/gisaid/trial/forecasting-time-length/nextstrain_clades/global/mlr/latest_results.json`

I confirmed that the additions to the JSON did not affect the generation of the current static images. 

I think we will have to update [nextstrain/forecasts-viz](https://github.com/nextstrain/forecasts-viz) to [parse](https://github.com/nextstrain/forecasts-viz/blob/main/src/lib/utils/parse.js#L79) and [graph](https://github.com/nextstrain/forecasts-viz/blob/05ac579a885d727bbaac4d9dbc63198b0d155d38/src/lib/components/SmallMultiple.js#L312) the forecast results.

**Comment by trvrb** [comment]
I was able to run this locally (pulling in latest changes to evofr through `nextstrain update`). Everything worked well and I'd think can be merged.
